jetty-http/src/main/java/org/eclipse/jetty/http/GZIPContentDecoder.java

```diff
@@ -46,83 +49,97 @@
         reset();
     }
 
-    /** Inflate compressed data from a buffer.
-     * 
-     * @param compressed Buffer containing compressed data.
-     * @return Buffer containing inflated data.
+    /**
+     * Inflates compressed data from a buffer.
+     *
+     * @param compressed the buffer containing compressed data.
+     * @return a buffer containing inflated data.
      */
     public ByteBuffer decode(ByteBuffer compressed)
     {
         decodeChunks(compressed);
-        if (BufferUtil.isEmpty(_inflated) || _state==State.CRC || _state==State.ISIZE )
-            return BufferUtil.EMPTY_BUFFER;
-        
-        ByteBuffer result = _inflated;
-        _inflated = null;
-        return result;
+
+        if (_inflateds.isEmpty())
+        {
+            if (BufferUtil.isEmpty(_inflated) || _state == State.CRC || _state == State.ISIZE)
+                return BufferUtil.EMPTY_BUFFER;
+            ByteBuffer result = _inflated;
+            _inflated = null;
+            return result;
+        }
+        else
+        {
+            _inflateds.add(_inflated);
+            _inflated = null;
+            int length = _inflateds.stream().mapToInt(Buffer::remaining).sum();
+            ByteBuffer result = acquire(length);
+            for (ByteBuffer buffer : _inflateds)
+            {
+                BufferUtil.append(result, buffer);
+                release(buffer);
+            }
+            _inflateds.clear();
+            return result;
+        }
     }
 
-    /** Called when a chunk of data is inflated.
-     * <p>The default implementation aggregates all the chunks 
+    /**
+     * <p>Called when a chunk of data is inflated.</p>
+     * <p>The default implementation aggregates all the chunks
      * into a single buffer returned from {@link #decode(ByteBuffer)}.
      * Derived implementations may choose to consume chunks individually
-     * and return false to prevent further inflation until a subsequent
+     * and return {@code true} to prevent further inflation until a subsequent
      * call to {@link #decode(ByteBuffer)} or {@link #decodeChunks(ByteBuffer)}.
-     *  
-     * @param chunk The inflated chunk of data
-     * @return False if inflating should continue, or True if the call 
-     * to {@link #decodeChunks(ByteBuffer)} or {@link #decode(ByteBuffer)} 
+     *
+     * @param chunk the inflated chunk of data
+     * @return false if inflating should continue, or true if the call
+     * to {@link #decodeChunks(ByteBuffer)} or {@link #decode(ByteBuffer)}
      * should return, allowing back pressure of compressed data.
      */
     protected boolean decodedChunk(ByteBuffer chunk)
     {
-        if (_inflated==null)
+        if (_inflated == null)
         {
-            _inflated=chunk;
+            _inflated = chunk;
         }
         else
         {
             int size = _inflated.remaining() + chunk.remaining();
-            if (size<=_inflated.capacity())
+            if (size <= _inflated.capacity())
```

Comment:
This logic looks wrong or at least confusing.
Just use BufferUtil!  This should be:
```
if (BufferUtil.space(_inflate) >= chunk.remaining())  
    BufferUtil.append(_inflated, chunk);
else
    ...
```

Reply:
Done.